File: packages/gavio-py/gavio/interceptors/guardrails/interceptor.py
"""GuardrailsInterceptor (F-QUA-01, F-QUA-02) — validate responses, act on failure.

An ExecutorPolicy so it can re-run the provider on failure. Runs every
validator against the response content; on failure it either raises, retries, or
warns. Records the outcome in ``ctx.guardrail_outcome`` for the audit trail.
"""

from __future__ import annotations

import logging

from ...context import InterceptorContext
from ...exceptions import GuardrailViolationError
from ...request import GavioRequest
from ...response import GavioResponse
from ...types import GuardrailOutcome
from ..chain import Executor
from ..executor import ExecutorPolicy
from .validator import OutputValidator

logger = logging.getLogger("gavio.guardrails")

_ACTIONS = {"error", "retry", "warn"}


class GuardrailsInterceptor(ExecutorPolicy):
    def __init__(
        self,
        validators: list[OutputValidator],
        on_failure: str = "error",
        max_retries: int = 2,
    ) -> None:
        if on_failure not in _ACTIONS:
            raise ValueError(f"on_failure must be one of {_ACTIONS}")
        self.validators = validators
        self.on_failure = on_failure
        self.max_retries = max_retries

    @property
    def name(self) -> str:
        return "guardrails"

    async def around(
        self,
        request: GavioRequest,
        ctx: InterceptorContext,
        call_next: Executor,
    ) -> GavioResponse:
        ctx.mark_fired(self.name)
        attempts = self.max_retries + 1 if self.on_failure == "retry" else 1
        response: GavioResponse | None = None
        failures: list[str] = []

        for attempt in range(attempts):
            response = await call_next(request)
            failures = [
                f"{v.name}: {result.reason}"
                for v in self.validators
                for result in [v.validate(response.content)]
                if not result.ok
            ]
            if not failures:
                ctx.guardrail_outcome = GuardrailOutcome.PASS.value
                return response
            logger.warning(
                "guardrails failed (attempt %d/%d): %s", attempt + 1, attempts, failures
            )

        ctx.guardrail_outcome = GuardrailOutcome.FAIL.value
        if self.on_failure == "warn":
            return response  # type: ignore[return-value]  # loop ran at least once
        raise GuardrailViolationError("; ".join(failures))

File: packages/gavio-js/src/interceptors/guardrails/interceptor.ts
/**
 * guardrails (F-QUA-01, F-QUA-02) — validate responses, act on failure.
 *
 * An ExecutorPolicy so it can re-run the provider on failure. Records the
 * outcome in ctx.guardrailOutcome for the audit trail.
 */

import type { InterceptorContext } from '../../context.js'
import { GuardrailViolationError } from '../../errors.js'
import type { GavioRequest } from '../../request.js'
import type { GavioResponse } from '../../response.js'
import type { Executor, ExecutorPolicy } from '../base.js'
import type { OutputValidator } from './validator.js'

export type OnFailure = 'error' | 'retry' | 'warn'

export interface GuardrailsOptions {
  validators: OutputValidator[]
  onFailure?: OnFailure
  maxRetries?: number
}

export function guardrails(options: GuardrailsOptions): ExecutorPolicy {
  const { validators, onFailure = 'error', maxRetries = 2 } = options
  return {
    name: 'guardrails',
    isExecutorPolicy: true,
    async around(
      request: GavioRequest,
      ctx: InterceptorContext,
      callNext: Executor,
    ): Promise<GavioResponse> {
      ctx.markFired('guardrails')
      const attempts = onFailure === 'retry' ? maxRetries + 1 : 1
      let response: GavioResponse | undefined
      let failures: string[] = []

      for (let attempt = 0; attempt < attempts; attempt++) {
        response = await callNext(request)
        failures = []
        for (const v of validators) {
          const result = v.validate(response.content)
          if (!result.ok) failures.push(`${v.name}: ${result.reason ?? ''}`)
        }
        if (failures.length === 0) {
          ctx.guardrailOutcome = 'PASS'
          return response
        }
      }

      ctx.guardrailOutcome = 'FAIL'
      if (onFailure === 'warn') return response as GavioResponse
      throw new GuardrailViolationError(failures.join('; '))
    },
  }
}
